worker/src/sidechain_storage/storage.rs

```diff
@@ -0,0 +1,887 @@
+		}
+		Ok(())
+	}
+
+	/// purges a shard and its block from the db storage
+	pub fn prune_shard_from_block_number(
+		&mut self,
+		shard: &ShardIdentifier,
+		block_number: BlockNumber,
+	) -> Result<()> {
+		if self.shards.contains(&shard) {
+			// get last block of shard
+			let last_block = match self.last_blocks.get(shard) {
+				Some(last_block) => last_block.clone(),
+				None => return Err(Error::LastBlockNotFound(*shard)),
+			};
+			if last_block.number == block_number {
+				// given block number is last block of chain - purge whole shard
+				self.purge_shard(shard)?;
+			} else {
+				// remove given block from db storage
+				let mut current_block_number = block_number;
+
+				// Remove blocks from db until no block anymore
+				while let Some(block_hash) = self.get_block_hash(&shard, current_block_number)? {
+					// Blockhash -> Signed Block
+					self.db.delete(block_hash.encode()).map_err(Error::OperationalError)?;
+					// (Shard, Block number) -> Blockhash
+					self.db
+						.delete((shard, current_block_number).encode())
+						.map_err(Error::OperationalError)?;
+					current_block_number -= 1;
+				}
+			}
+		}
+		Ok(())
+	}
+
+	/// prunes all shards except for the newest blocks (according to blocknumber)
+	pub fn prune_shards(&mut self, blocks_to_keep: BlockNumber) {
+		for shard in self.shards().clone() {
+			// get last block:
+			if let Some(last_block) = self.last_block_of_shard(&shard) {
+				let threshold_block = last_block.number - blocks_to_keep;
+				if let Err(e) = self.prune_shard_from_block_number(&shard, threshold_block) {
+					error!("Could not purge shard {:?} due to {:?}", shard, e);
+				}
+			} else {
+				error!("Last block not found in shard {:?}", shard);
+			}
+		}
+	}
+}
+
+/// implementations of helper functions, not meant for pub use
+impl SidechainStorage {
+	/// gets the previous block of given shard and block number, if there is one
+	fn get_previous_block(
+		&self,
+		shard: &ShardIdentifier,
+		current_block_number: BlockNumber,
+	) -> Option<LastSidechainBlock> {
+		let prev_block_number = current_block_number - 1;
+		if let Some(block_hash_encoded) = self.db.get((shard, prev_block_number).encode()).unwrap()
+		{
+			let block_hash = H256::decode(&mut block_hash_encoded.as_slice()).unwrap();
+			Some(LastSidechainBlock { hash: block_hash, number: prev_block_number })
+		} else {
+			None
+		}
+	}
+	/// reads shards from DB
+	fn load_shards_from_db(db: &DB) -> Result<Vec<ShardIdentifier>> {
+		match db.get(STORED_SHARDS_KEY.encode()).map_err(Error::OperationalError)? {
+			Some(shards) => Ok(Vec::<ShardIdentifier>::decode(&mut shards.as_slice())
+				.map_err(Error::DecodeError)?),
+			None => Ok(vec![]),
+		}
+	}
+
+	/// reads last block from DB
+	fn load_last_block_from_db(
+		db: &DB,
+		shard: &ShardIdentifier,
+	) -> Result<Option<LastSidechainBlock>> {
+		match db.get((LAST_BLOCK_KEY, *shard).encode()).map_err(Error::OperationalError)? {
+			Some(last_block_encoded) => Ok(Some(
```

Comment:
here you could aso use the extracted: `fn get<V:Decode>(key: Vec<u8>) -> Result<Option<V>>`

Reply:
Good thinking, thanks. Implemented in https://github.com/integritee-network/worker/pull/380/commits/9de273401f0ebef69e9927a3b355013cdd603578